### PR TITLE
[SPARK-18932][SQL] Support partial aggregation for collect_set/collect_list

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Currently collect_set/collect_list aggregation expression don't support partial aggregation. This patch is to enable partial aggregation for them.

## How was this patch tested?

Jenkins tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala [hunk 1]
```diff
@@ -33,10 +33,9 @@ import org.apache.spark.util.collection.OpenHashMap
  * The Percentile aggregate function returns the exact percentile(s) of numeric column `expr` at
  * the given percentage(s) with value range in [0.0, 1.0].
  *
- * The operator is bound to the slower sort based aggregation path because the number of elements
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala]: `TypedImperativeAggregate` is't bound to sort based aggregation. `ObjectHashAggregateExec` supports `TypedImperativeAggregate` in hash based aggregation now.

**Comment by rxin** [comment]
Why would we even want to support partial aggregation for collect_list?

**Comment by rxin** [comment]
@marmbrus I see that you filed the ticket. We can make this work with partial aggregation, but we shouldn't run them this way. Maybe it's time to define something as partial aggregateable vs should be partially aggregated.

**Comment by viirya** [comment]
If we can have some definition for something should be partially aggregated and something not, that would be great.

**Comment by viirya** [comment]
Just for reference. Hive's collect_list and collect_set seems supporting partial aggregation. Of course we don't necessarily follow Hive's.

https://github.com/apache/hive/blob/ff67cdda1c538dc65087878eeba3e165cf3230f4/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java

**Comment by marmbrus** [comment]
Is there a problem with running partial aggregation for `collect_list`?  Even if there isn't an improvement, at least now you don't disable partial aggregation for other functions that are also being computed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 2]
```diff
@@ -44,39 +43,52 @@ abstract class Collect extends ImperativeAggregate {
...
-    buffer.clear()
-  }
-
-  override def update(b: InternalRow, input: InternalRow): Unit = {
-    // Do not allow null values. We follow the semantics of Hive's collect_list/collect_set here.
-    // See: org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMkCollectionEvaluator
-    val value = child.eval(input)
-    if (value != null) {
-      buffer += value
+  protected def _serialize(obj: Iterable[Any]): Array[Byte] = {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: Should we generalize this code? It is also part the Percentile class.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala [hunk 3]
```diff
@@ -33,10 +33,9 @@ import org.apache.spark.util.collection.OpenHashMap
...
- * and their partial order cannot be determined in advance. Therefore we have to store all the
- * elements in memory, and that too many elements can cause GC paused and eventually OutOfMemory
- * Errors.
+ * Because the number of elements and their partial order cannot be determined in advance.
+ * Therefore we have to store all the elements in memory, and that too many elements can cause GC
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala]: Could you fix the language? `..., and that too...` That sentence is weird. This is probably on me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 4]
```diff
@@ -44,39 +43,52 @@ abstract class Collect extends ImperativeAggregate {
...
     }
   }
 
-  override def merge(buffer: InternalRow, input: InternalRow): Unit = {
-    sys.error("Collect cannot be used in partial aggregations.")
-  }
-
-  override def eval(input: InternalRow): Any = {
-    new GenericArrayData(buffer.toArray)
+  protected def _deserialize(bytes: Array[Byte], updater: (Any) => Unit): Unit = {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: Same here.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: You could also pass in a `Growable[]` instead of an anonymous function.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 5]
```diff
@@ -29,12 +30,10 @@ import org.apache.spark.sql.types._
...
  *
- * The operator is bound to the slower sort based aggregation path because the number of
- * elements (and their memory usage) can not be determined in advance. This also means that the
- * collected elements are stored on heap, and that too many elements can cause GC pauses and
- * eventually Out of Memory Errors.
+ * We have to store all the collected elements in memory, and that too many elements can cause GC
+ * paused and eventually OutOfMemory Errors.
  */
-abstract class Collect extends ImperativeAggregate {
+abstract class Collect[T] extends TypedImperativeAggregate[T] {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: Why [T] and not T <: Growable[Any] with Iterable[Any]?

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: If we can do that, we can make the PR smaller.

**Comment by hvanhovell** [comment]
@viirya I left a few comments on the PR. Lets wait for some consensus before pushing ahead.

**Comment by rxin** [comment]
sounds good.

**Comment by viirya** [comment]
@hvanhovell Got it. Thanks for review.

**Comment by cloud-fan** [comment]
I think we should implement collect with `TypedImperativeAggregate` and remove the `supportsPartual` flag, @viirya do you have time to work on it?

**Comment by viirya** [comment]
@cloud-fan I already did this in current PR. I am waiting for some consensus so I can push ahead to address the review comments from @hvanhovell.

**Comment by cloud-fan** [comment]
I think we should support partial aggregation for collect, in the future we can define an interface to declare "should partial aggregate", so that a single collect function won't cause a 2-phase aggregation.

I think they are orthogonal and we can go ahead.

**Comment by marmbrus** [comment]
+1 I think we can move forward.

**Comment by viirya** [comment]
@cloud-fan @marmbrus ok. I will first address @hvanhovell's comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: `Percentile` takes `OpenHashMap`, which has a bit different behavior than `Growable[]`.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: We can generalize this to cover `Percentile`, but we may not use `T <: Growable[Any] with Iterable[Any]` but `T <: Iterable[Any]` then.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: Also, the serialized format (child.dataType, LongType) of `Percentile` is different than `CollectList` and `CollectSet` (i.e., child.dataType).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala (hunk 5 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: yeah, that's good idea.

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 6]
```diff
@@ -44,39 +44,78 @@ abstract class Collect extends ImperativeAggregate {
...
-  override def merge(buffer: InternalRow, input: InternalRow): Unit = {
-    sys.error("Collect cannot be used in partial aggregations.")
+  override def eval(buffer: T): Any = {
+    generateOutput(buffer)
   }
 
-  override def eval(input: InternalRow): Any = {
-    new GenericArrayData(buffer.toArray)
+  protected def _serialize(obj: Iterable[Any]): Array[Byte] = {
+    val buffer = new Array[Byte](4 << 10)  // 4K
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: shall we create `new GenericArrayData(buffer.toArray)` and do unsafe projection?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: I think the effect is the same. Any benefit by doing this?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: I just tried it locally. Looks like the code can be more compact.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: @cloud-fan I did the change, is this looks good for you?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 7]
```diff
@@ -44,39 +44,48 @@ abstract class Collect extends ImperativeAggregate {
...
   // Both `CollectList` and `CollectSet` are non-deterministic since their results depend on the
   // actual order of input rows.
   override def deterministic: Boolean = false
 
-  protected[this] val buffer: Growable[Any] with Iterable[Any]
-
-  override def initialize(b: InternalRow): Unit = {
-    buffer.clear()
+  private def generateOutput(results: Iterable[Any]): Any = {
+    if (results.isEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: is it a behavior change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 8]
```diff
@@ -44,39 +44,48 @@ abstract class Collect extends ImperativeAggregate {
...
+      null
+    } else {
+      new GenericArrayData(results.toArray)
+    }
   }
 
-  override def update(b: InternalRow, input: InternalRow): Unit = {
+  override def update(buffer: T, input: InternalRow): T = {
+    val value = child.eval(input).asInstanceOf[Any]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: unnecessary `asInstanceOf`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 9]
```diff
@@ -44,39 +44,48 @@ abstract class Collect extends ImperativeAggregate {
...
+    buffer ++= other
   }
 
-  override def eval(input: InternalRow): Any = {
-    new GenericArrayData(buffer.toArray)
+  override def eval(buffer: T): Any = {
+    generateOutput(buffer)
+  }
+
+  protected def _serialize(obj: Iterable[Any]): Array[Byte] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: can't we just define `serialize` here?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: `override def serialize(buffer: T)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 10]
```diff
@@ -44,39 +44,48 @@ abstract class Collect extends ImperativeAggregate {
...
+    val array = new GenericArrayData(obj.toArray)
+    val projection = UnsafeProjection.create(
+      Array[DataType](ArrayType(elementType = child.dataType, containsNull = false)))
+    projection.apply(InternalRow.apply(array)).getBytes()
+  }
+
+  protected def _deserialize(bytes: Array[Byte], buffer: T): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: same here
```
override def deserialize(bytes: ArrayByte): Unit = {
  val buffer = createAggregationBuffer()
  ...
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 11]
```diff
@@ -44,39 +44,48 @@ abstract class Collect extends ImperativeAggregate {
...
 
-  override def eval(input: InternalRow): Any = {
-    new GenericArrayData(buffer.toArray)
+  override def eval(buffer: T): Any = {
+    generateOutput(buffer)
+  }
+
+  protected def _serialize(obj: Iterable[Any]): Array[Byte] = {
+    val array = new GenericArrayData(obj.toArray)
+    val projection = UnsafeProjection.create(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: make this project a member variable, then we don't need to create it everytime we call `serialize`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala [hunk 12]
```diff
@@ -63,7 +63,7 @@ class RewriteDistinctAggregatesSuite extends PlanTest {
     val input = testRelation
       .groupBy('a, 'd)(
         countDistinct('e, 'c).as('agg1),
-        CollectSet('b).toAggregateExpression().as('agg2))
+        DummpAgg('b).toAggregateExpression().as('agg2))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: shall we just remove the `supportsPartial` and this test?

Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: i was thought about removing it. however, `RewriteDistinctAggregates` rule has logic for this case. this test is the only one for testing this logic. so i didn't remove it in the end.

Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: we can remove that logic in `RewriteDistinctAggregates` too

Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: We should remove the support for non-partial aggregation in that case.

Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: because we don't have non-partial aggregation function now?

Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: If we remove the logic from RewriteDistinctAggregates, we need to make sure that non-partial aggregates do not exist anymore. Lets do this in a follow-up.

Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: ok. make sense to me.

Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: I will only remove this test in this pr.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala (hunk 7 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: fixed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala [hunk 13]
```diff
@@ -16,16 +16,16 @@
...
 import org.apache.spark.sql.catalyst.dsl.expressions._
 import org.apache.spark.sql.catalyst.dsl.plans._
-import org.apache.spark.sql.catalyst.expressions.{If, Literal}
-import org.apache.spark.sql.catalyst.expressions.aggregate.{CollectSet, Count}
+import org.apache.spark.sql.catalyst.expressions.{Expression, If, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{CollectSet, Count, ImperativeAggregate, TypedImperativeAggregate}
 import org.apache.spark.sql.catalyst.plans.PlanTest
 import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Expand, LocalRelation, LogicalPlan}
-import org.apache.spark.sql.types.{IntegerType, StringType}
+import org.apache.spark.sql.types.{DataType, IntegerType, StringType}
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregatesSuite.scala]: please revert unnecessary changes in `import`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 14]
```diff
@@ -88,19 +92,19 @@ abstract class Collect extends ImperativeAggregate {
 case class CollectList(
     child: Expression,
     mutableAggBufferOffset: Int = 0,
-    inputAggBufferOffset: Int = 0) extends Collect {
+    inputAggBufferOffset: Int = 0) extends Collect[ArrayBuffer[Any]] {
 
   def this(child: Expression) = this(child, 0, 0)
 
-  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): ImperativeAggregate =
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): CollectList =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 15]
```diff
@@ -88,19 +92,19 @@ abstract class Collect extends ImperativeAggregate {
...
     copy(mutableAggBufferOffset = newMutableAggBufferOffset)
 
   override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ImperativeAggregate =
     copy(inputAggBufferOffset = newInputAggBufferOffset)
 
-  override def prettyName: String = "collect_list"
+  override def createAggregationBuffer(): ArrayBuffer[Any] = new ArrayBuffer[Any]()
 
-  override protected[this] val buffer: mutable.ArrayBuffer[Any] = mutable.ArrayBuffer.empty
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: `mutable.ArrayBuffer.empty` looks better.

**Comment by cloud-fan** [comment]
LGTM

**Comment by viirya** [comment]
retest this please.

**Comment by hvanhovell** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by viirya** [comment]
Thank you @cloud-fan @hvanhovell @rxin @marmbrus
